doc/implicit_conversion.rdoc

```diff
@@ -0,0 +1,187 @@
+== Implicit Conversion
+
+Some Ruby methods accept one or more objects
+that can be either:
+* <i>Of a given class</i>, and so accepted as is.
+* <i>Implicitly convertible to that class</i>, in which case
+  the called method converts the object.
+
+For each of the relevant classes, the conversion is done by calling
+a specific conversion method:
+
+* Array: +to_ary+
+* Hash: +to_hash+
+* Integer: +to_int+
+* String: +to_str+
+
+=== Array-Convertible Objects
+
+An <i>Array-convertible object</i> is an object that:
+* Has instance method +to_ary+.
+* The method accepts no arguments.
+* The method returns an object +obj+ for which <tt>obj.kind_of?(Array)</tt> returns +true+.
+
+The examples in this section use method <tt>Array#replace</tt>,
+which accepts an Array-convertible argument.
+
+This class is Array-convertible:
+
+    class ArrayConvertible
+      def to_ary
+        [:foo, 'bar', baz = 2]
+      end
+    end
+    a = []
+    a.replace(ArrayConvertible.new) # => [:foo, "bar", 2]
+
+This class is not Array-convertible (no +to_ary+ method):
+
+    class NoToAry; end
+    a = []
+    a.replace(NoToAry.new) # Raises TypeError (no implicit conversion of NoToAry into Array)
+
+This class is not Array-convertible (method +to_ary+ takes arguments):
+
+    class ToAryTakesArguments
+      def to_ary(x)
+        [:foo, 'bar', baz = 2]
+      end
+    end
+    a = []
+    a.replace(ToAryTakesArguments.new) # Raises ArgumentError (wrong number of arguments (given 0, expected 1))
+
+This class is not Array-convertible (method +to_ary+ returns non-Array):
+
+    class ToAryReturnsNonArray
+      def to_ary
+        :foo
+      end
+    end
+    a = []
+    a.replace(ToAryReturnsNonArray.new) # Raises TypeError (can't convert ToAryReturnsNonArray to Array (ToAryReturnsNonArray#to_ary gives Symbol))
```

Comment:
My preference is to avoid trailing comments. Put the comment on the line before. Also, this comment could probably be written without the nested brackets.